### DISCUSSION TITLE
Feature Request: optimize queries with list_aggregate and unnest on same array

### DISCUSSION THREAD

**Discussion by rpbouman** [duckdb/duckdb#15314] [Ideas]
There appears to be an opportunity to optimize list_aggregate that is currently missed.

Example query:

```
with data as (
    SELECT range(100000) as arr
), cte1 as (
  SELECT list_aggregate(arr, 'COUNT') as arr_count
  ,      list_aggregate(arr, 'AVG')   as arr_avg
  ,      unnest(arr)
  FROM data
)
select * 
from cte1;
```

On my laptop the query above takes 71 seconds. 
The equivalent query:

```
with data as (
    SELECT range(100000) as arr
), cte1 as (
  SELECT list_aggregate(arr, 'COUNT') as arr_count
  ,      list_aggregate(arr, 'AVG')   as arr_avg
  ,      arr
  FROM data
), cte2 as (
  SELECT arr_count
  ,      arr_avg
  ,      unnest(arr)
  FROM cte1
)
select * 
from cte2; 
```

takes 0.313 seconds

In both cases, the result is driven by the unnesting operation on the array. The first query probably takes longer because the list_aggregate expressions are computed again for each row generated by the unnesting operation.

My use case is that the user interactively explores the dataset. They might start by selecting some list_aggregates to profile the array, and then unnest to drill down to the details. Unfortunately, for the real-world datasets I encounter this is issue, the unnesting operation in combination with the list_aggregates slows the app down to a crawl, and sometimes makes DuckDB run out of memory.

To be sure, there is a workaround as I can myself rewrite the query to include a separate stage for unnesting the data after calculating the list_aggregate results. But it would be extremely nice if DuckDB could detect this and return the cached result of the list_aggregate expressions.

**Comment by soerenwolfers** [comment]
If I were you, I'd file this as a proper issue rather than a feature request, and I'd include whether this applies more generally to 

```sql 
SELECT f(x), unnest(y)
```
where `x` can be any type and `f` any function, or whether it requires that `x=y` and that `f` be a list aggregate rather than any other list function.

**Reply by rpbouman** [reply]
Hi Soeren, thanks for chiming in!

> If I were you, I'd file this as a proper issue rather than a feature request, 

I considered that but wanted to gauge if this idea is too niche or is worthy of an issue. I think there are quite a bunch of issues already that should probably take priority and I didn't want to add to the pile right away.

>  I'd include whether this applies more generally to

Actually, I did consider that, but then it occurred to me that it might not be so straightforward to effectively optimize that in the general case, and that this case really is special in a way. My reasoning is:

1) the SELECT that executes f(x) might not be a single row, but could be on a large number of rows, at which point the cost of memorizing all function results might not be so trivial as compared to the cost of calculating the function - especially if the function arguments are dependent upon the current row data.
2) in the case where the array aggregator and the unnesting happen in the same context, and on the same array value, you already know you need to unfold the array. It felt natural to want to weave the aggregation computation into that process

> requires that x=y and that f be a list aggregate rather than any other list function.

I guess that's a fair point! Most list functions will need to loop the list and this could be combined with the unnesting process.

But at any rate, I don't really know about DuckDB internals, so I don't want to assume too much about what would be a sensible optimization.

**Reply by soerenwolfers** [reply]
> I considered that but wanted to gauge if this idea is too niche or is worthy of an issue. I think there are quite a bunch of issues already that should probably take priority and I didn't want to add to the pile right away.

You can let the developers decide. Opening an issue certainly doesn't dictate their prioritization. Especially if there is a simple rewrite that's 210x faster, I think it's fair to ask whether the optimizer should have noticed that itself. They can always close the issue or put it on review. 

> Most list functions will need to loop the list and this could be combined with the unnesting process.

I don't think the looping has anything do do with it. Any potentially slow function should only be executed once if there is an option to.

**Comment by rpbouman** [comment]
After checking with @Mytherin I acted on @soerenwolfers' suggestion and made an issue for it: https://github.com/duckdb/duckdb/issues/15341. 

Closing discussion as it is now an issue.
